### PR TITLE
SSL/TLS: Parse CertificateRequest message

### PR DESCRIPTION
**Pull request by 0xxon**
This commit introduces parsing of the CertificateRequest message in the TLS handshake. It introduces a new event ssl_certificate_request, as well as a new function parse_distinguished_name, which can be used to parse part of the ssl_certificate_request event parameters.

This commit also introduces a new policy script, which appends information about the CAs a TLS server requests in the CertificateRequest message, if it sends it.

Please note that merging this also requires merging the updated branches in both external testing repos.

### CODE REVIEW

File: testing/btest/Baseline/scripts.policy.protocols.ssl.certificate-request-info/ssl.log [hunk 1]
```diff
@@ -0,0 +1,11 @@
+### BTest baseline data generated by btest-diff. Do not edit. Use "btest -U/-u" to update. Requires BTest >= 0.63.
+#separator \x09
+#set_separator	,
+#empty_field	(empty)
+#unset_field	-
+#path	ssl
+#open XXXX-XX-XX-XX-XX-XX
+#fields	ts	uid	id.orig_h	id.orig_p	id.resp_h	id.resp_p	version	cipher	curve	server_name	resumed	last_alert	next_protocol	established	ssl_history	cert_chain_fps	client_cert_chain_fps	sni_matches_cert	requested_client_certificate_authorities
+#types	time	string	addr	port	addr	port	string	string	string	string	bool	string	string	bool	string	vector[string]	vector[string]	bool	vector[string]
+XXXXXXXXXX.XXXXXX	CHhAvVGS1DHFjwGM9	fd42:496a:d659:bb85::1	52464	fd42:496a:d659:bb85:216:3eff:fe6a:a257	3000	TLSv12	TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384	x25519	-	F	handshake_failure	-	F	CsxkrnXGIl	0a171ee771a26530c650fe8b8a6bf205177bfb64fbb3e5303ba348c13ffc7dfa,c628dd5aae1f216da6ce4f8f914fb7141c2b0afd3522cce5900bcc4840657bfd	(empty)	-	O=SomeOrg\x2cL=Somewhere\x2cST=Some-State\x2cC=US
```
Comment (awelzel) [inline testing/btest/Baseline/scripts.policy.protocols.ssl.certificate-request-info/ssl.log:10]: `\x2c`'s are the ASCII writer escaping the `,`? With JSON that won't happen?

File: scripts/policy/protocols/ssl/certificate-request-info.zeek [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+	requested_client_certificate_authorities: vector of string &optional &log;
+};
+
+event ssl_certificate_request(c: connection, is_client: bool, certificate_types: index_vec, supported_signature_algorithms: SSL::SignatureAndHashAlgorithm, certificate_authorities: string_vec)
+	{
+	if ( is_client )
+		return;
+
+	local out: vector of string = vector();
+	for ( i in certificate_authorities )
```
Comment (awelzel) [inline scripts/policy/protocols/ssl/certificate-request-info.zeek:19]: If you wanted to shorten a bit.


```suggestion
	for ( _, ca in certificate_authorities )
	    out += parse_distinguished_name(ca);
```

File: testing/btest/scripts/base/protocols/ssl/certificate_request.zeek [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+# @TEST-EXEC: btest-diff out
+
+@load base/protocols/ssl
+@load base/protocols/mysql
+
+event ssl_certificate_request(c: connection, is_client: bool, certificate_types: index_vec, supported_signature_algorithms: SSL::SignatureAndHashAlgorithm, certificate_authorities: string_vec)
+	{
+	print certificate_types;
+	print supported_signature_algorithms;
+	for ( i in certificate_authorities )
```
Comment (awelzel) [inline testing/btest/scripts/base/protocols/ssl/certificate_request.zeek:16]: Could do `for ( _, ca in certificate_authorities)` if you wanted for a bit shortening.

**Review by awelzel — APPROVED** [review]
Nits only.

File: testing/btest/Baseline/scripts.policy.protocols.ssl.certificate-request-info/ssl.log (hunk 1 above)
Comment (0xxon) [reply testing/btest/Baseline/scripts.policy.protocols.ssl.certificate-request-info/ssl.log:10]: yup.

File: scripts/policy/protocols/ssl/certificate-request-info.zeek (hunk 2 above)
Comment (0xxon) [reply scripts/policy/protocols/ssl/certificate-request-info.zeek:19]: oh, right, I totally forgot that syntax :)

File: src/analyzer/protocol/ssl/tls-handshake-analyzer.pac [hunk 4]
```diff
@@ -626,6 +626,47 @@ refine connection Handshake_Conn += {
...
+		%{
+		if ( ! ssl_certificate_request )
+			return true;
+
+		auto ctlist = zeek::make_intrusive<zeek::VectorVal>(zeek::id::find_type<zeek::VectorType>("index_vec"));
+		auto ctypes = ${req.certificate_types};
+
+		if ( ctypes )
+			for ( unsigned int i = 0; i < ctypes->size(); ++i)
+				ctlist->Assign(i, zeek::val_mgr->Count((*ctypes)[i]));
```
[the remark this answers is not in the document]
Comment (0xxon) [reply src/analyzer/protocol/ssl/tls-handshake-analyzer.pac:639]: This one is a bit of a stylistic preference - I kind of prefer the explicitness of the equivalence of the positions :)

**Pull request merged**